File: cpp/Backtracking/L_subset.cpp
#include<bits/stdc++.h>
using namespace std;

vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> curr;
        backtrack(nums, res, curr, 0);
        return res;
    }
    void backtrack(vector<int>& nums, vector<vector<int>>& res,vector<int>& curr, int start ) {
        res.push_back(curr);
        for(int i=start;i<nums.size();i++) {
            curr.push_back(nums[i]);
            backtrack(nums, res, curr, i+1);
            curr.pop_back();
        }
    }

int main() {

}

File: Python/Backtracking/L_subset.py
def subsets(nums):
        # res = [[]]
        # for num in sorted(nums):
        #     res += [item+[num] for item in res]
        # return res
        res = []
        def backtrack(n,c):
            res.append(c)
            for i in range(len(n)):
                backtrack(n[i+1:], c+[n[i]])
                
        if nums:
            backtrack(nums,[])
        return res



def subsets(nums):
        res = []
        def backtrack(i, subset):
            if i == len(nums):
                res.append(subset.copy())
                return
            subset.append(nums[i])
            backtrack(i+1, subset)
            subset.pop()
            backtrack(i+1, subset)
            
        if nums:
            backtrack(0,[])
        return res